### DISCUSSION TITLE
Expanding the callback

### DISCUSSION THREAD

**Discussion by rolfiepolfie** [mido/mido#547] [Q&A] [answered]
So you all know about this from the manual:

```py
port = mido.open_input(callback= somefunction)
```
or
```py
port.callback = somefunction
```

and your callback function (the `somefunction` is here called `callback`):
```py
def callback(msg):
  # do something clever 
```

However I am having lots of devices using ports evoking different callbacks. 
I would like them all to be connected to a common callback in which I also can add an extra parameter for extra info, like which device 
was evoking the callback.
```py
def def callback(msg, extra_parameter):
  if extra_parameter.deviceName == 'ROLAND':
    # do something explicit for the Roland device 
       
  if extra_parameter.deviceName == 'KAWAI':
    # do something explicit for the Kawai device 
```

I have tried to extract info from the `def callback` to see if I can extract info like the `device name` used  when opening the port, but without luck
Also I have starting looking at the `rtmidi.py` file , but not looking forward to all the torture
Are there any way sort of simpler way to attack this? , thanks!

**Comment by rdoursenaud** [comment, marked as answer]
You may want to have a look at my answer to a previous discussion: https://github.com/orgs/mido/discussions/389#discussioncomment-3766289

**Reply by rolfiepolfie** [reply]
Thanks , your solution is insane (in a good way) , insanely good

**Reply by rolfiepolfie** [reply]
I have made a collection with 2 of your midi-in device classes for test  (called Inport in my case), after some fiddeling I found something that works.
The code below shows using (newports.callback) the names of my devices like  'Live Logic MIDI Controller 0' and 'MPKMini2 1' when I play on them.

Using a common callback for both the Inports in the same class. 
However I feel bad for calling one class from another class without any further structure (indicating with an arrow),
 could you if you got time provide some criticism on this code below?, thanks

**Comment by rolfiepolfie** [comment]
```py
import mido

class Inport():

    _name:  str                # The port/device name
    _port:  mido.ports.BaseInput  

    def __init__(self, name: str):
        self._name = name
        self._port = mido.open_input(name)
        self._port.callback = self.callback
        

    def callback(self, msg):

        newports.callback(self, msg) # <---------------- feeling bad here
        
    def getname(self): return self._name    


    def getCallback(self): return self._port.callback

    def poll(self):
        self._port.poll()
        
    def setCallback(self, cb):
        self._port.callback = cb
                
    def close(self):
        self._port.callback = None
        self._port.close()
        
    def __del__(self):
       self._port.callback = None
       self._port.close()
       

class newports():
    inport1: Inport
    inport2: Inport
        
    def __init__(self):
        
        self.inport1 = Inport('Live Logic MIDI Controller 0')
        #self.inport1.setCallback(self.callback)

        self.inport2 = Inport('MPKmini2 1')
        #self.inport2.setCallback(self.callback)
                    
    def callback(self, message):            
        print(self._name)

    def close(self):        
        self.inport1.close()
        self.inport2.close()
        
    def startMsgLoop(self):
        # message loop
        while True:
            self.inport1.poll()
            self.inport2.poll()
            
       
tp=newports()
tp.startMsgLoop() 
```

**Reply by rdoursenaud** [reply]
You seem to be overcomplicating things.

First, since you're using a `callback`, you don't have to `poll` so you can get rid of the whole `startMsgLoop` thing.
The principle of the `callback` is that the backend (`rtmidi`) calls the function in a new thread each time a message is incoming.
It replaces polling entirely.

Next, you don't need to define the `close()` `method` unless you plan on calling it.
It's automatically called when the object is destroyed in the `__del__()` method.
This means that port automatically closes once you reach the end of the program or it is interrupted.
You can also force it by simply getting rid of the object: `del some_input`

Finally, I don't believe you need 2 `classes`.
What you need is two `objects` instantiating the same `class`.
Modifying the code I originally linked to, here's how I'd handle your use case:
```py
# myscript.py

import sys
import mido

DEBUG = True  # Change to False for production
LIVE_CONTROLLER_PORT_NAME = 'Live Logic MIDI Controller 0'
MPK_MINI_PORT_NAME = 'MPKmini2 1'

def handle_live_logic_controller(message: mido.Message):
    """Handles MIDI messages received from the Live Logic Controller"""
    raise NotImplementedError  # TODO: replace by doing something with the messages

def handle_mpk_mini(message: mido.Message):
    """Handles MIDI messages received from the MPK Mini"""
    raise NotImplementedError  # TODO: replace by doing something with the messages

class CustomInport:  # I use a different name here just to avoid confusion with the mido provided class.
    """A thin wrapper around mido input port"""
    name: str  # The port name
    port: mido.ports.BaseInput   # The port itself

    def __init__(self, name: str) -> None:
        self.name = name
        self.port = mido.open_input(self.name)
        self.port.callback = self.receive_callback

    def __del__(self) -> None:
        self.port.callback = None
        self.port.close()

    def receive_callback(self, message: mido.Message) -> None:
        """A generic callback"""
        # Do your thing here. You have access to self.name and self.port to know which port you are.
        if DEBUG:
            print(f'Message from {self.name} received: {message}')
        if self.name == LIVE_CONTROLLER_PORT_NAME:
            handle_live_logic_controller(message)
        elif self.name = MPK_MINI_PORT_NAME:
            handle_mpk_mini(message)
        return

###
# Usage
###

if __name__ == '__main__':

    # Open the ports and automatically set the callback
    live_controller_input = CustomInport(LIVE_CONTROLLER_PORT_NAME)
    mini_mpk_input = CustomInport(MPK_MINI_PORT_NAME)
  
    # Main loop
    While True:
        pass  # FIXME: define a condition to terminate the program
        # As long as the loop is going:
        # handle_logic_controller() & handle_mpk_mini() will be called each time a new message arrives.
        # You can easily check by printing the message in the `handle_` functions code for example.
  
    sys.exit(0)
```

**Reply by rdoursenaud** [reply]
Oh and in Python we usually don't use protected members and getters and setters unless absolutely needed.
We simply declare the properties and use them!
Look like you learned java first ;)

**Reply by rdoursenaud** [reply]
Edited the code for slight corrections after review O\

**Reply by rdoursenaud** [reply]
I've added the `handle_xxx()` methods for clarity but you can also do all your processing inside the `receive_callback()` depending on what you're trying to achieve.

**Reply by rolfiepolfie** [reply]
Yes you are correct, I have a tendency to overcomplicate things, also I have some Java/C++ background, we learned these in engineering school some years ago. I am aware about some OO-concepts, but they do not always translate in Pythonic ways, but we're getting there.

I also have a tendency to overuse the number of classes as I prepare for the future, this thing I work on has grown on me and more features are added, but also have resulted in several restarts with a complete rewrite. 

Your example demonstrates the simplicity and purity/readability we should strive for, thank you for the effort.

**Reply by rdoursenaud** [reply]
Thanks :)

Java has been my first object-oriented language and it left some scars I'm trying to atone by using Python ;)

If you haven't yet, you'd want to read the [Zen of Python](https://peps.python.org/pep-0020/).

Getters & setters are not needed in Python because we have the `@property` decorator to add the feature after the fact, if we need to, without changing the public API of the class/objects.

I appreciate that you noted the effort on readability & simplicity.
It's what I always strive for, in that order!
The Python syntax makes it even easier.

Feel free to ping me if you need code review, I very much enjoy these small studies!
I wish you success in your endeavor.

**Reply by rolfiepolfie** [reply]
yes thanks, I plan several request in the near future,  I am now rewriting everything in accordance to your input. 
In the next version I plan to make (starting off with something resemble your latter example) only making it  more "dynamic", that it will notify itself about devices being connected/disconnected and "adapt" to it automatically. 

For example when the user add a device (insert a USB cable) it will prompt the user how it will be utilized or something along  these lines, but at least involve the device in some kind of message dispatching by default so the user can utilize its messages at request if needed in a live session. 

Also I plan a "learning" feature, for example "please press a button for engaging the: minor-function" (this is a functionality converts a chord to a minor chord, for example: C6 -> C6 minor) 

If you should ask, I am a piano-player and my app builds chords  based on what I play on the piano-keyboard and and other midi-switches on the floor, these chords I will then send to for example software-guitars that enables me to play two instruments at once, having a guitar accompanying me.

I also send midi-messages to these software instruments configure them in a live session (midi-sliders). The keyword is "live" which it must adapt to several types of inputs very fast with a minimum of interaction. If this thing works conceptually  after some tuning I may try to sell it for a small group of live musicians.
